feat(integrations): add an integration with `ultralytics` YOLOv8 library

parambharat opened the pull request:
The minimal integration addresses the yolov8 version of the library. 
We log metrics and model artifacts and sample media produced by the library in the first version of this integration.


Description
-----------
This PR adds a minimal wandb integration to the [ultralytics](https://pypi.org/project/ultralytics/) library. Specifically, we tackle logging of metrics, model artifacts and media related to YOLOv8 models.

Testing
-------
How was this PR tested?
The following colab [notebook](https://colab.research.google.com/drive/1SO2lr2pFmhojtR0N9rifhCXkgislD5R6#scrollTo=lIf5QKbNplcS)
 shows how to use the logger for different tasks in the library. the following as example logged runs for each task in the notebook.
1. [Classification](https://wandb.ai/parambharat/YOLOv8/runs/xflis8ks)
2. [Object Detection](https://wandb.ai/parambharat/YOLOv8/runs/upyrv2mr)
2. [Image Segmentation](https://wandb.ai/parambharat/YOLOv8/runs/7zwesv5p)

Checklist
-------

- [X] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

Commit:
feat(integrations): add an integration with `ultralytics` YOLOv8 library

The minimal integration addresses the yolov8 version of the library.
We log metrics and model artifacts and sample media produced by the library in the first version of this integration.

ayulockin commented:
~@parambharat the ideal usage should be `from wandb.yolov8 import WandbLogger`. Following the dir structure of `wandb/wandb`, we should have a `wandb/yolov8` dir with just a `__init__.py` file. For example check out the Keras one: https://github.com/wandb/wandb/blob/main/wandb/keras/__init__.py~

ayulockin commented on wandb/integration/yolov8/yolov8.py:
````diff
@@ -0,0 +1,201 @@
+    A YOLO model wrapper that tracks metrics, and logs models to Weights & Biases.
+    Usage:
+    ```python
+    from wandb.integration.yolov8 import WandbLogger
````
commented on usage above.

ayulockin commented on wandb/integration/yolov8/yolov8.py:
```diff
@@ -0,0 +1,201 @@
+                    project=trainer.args.project or self.project or "YOLOv8",
+                    tags=["YOLOv8"] or self.tags,
+                    config=vars(trainer.args),
+                    resume="allow" or self.resume,
```
allow `kwargs` for additional `wandb.init` arguments if passed by the user?

ayulockin commented on wandb/integration/yolov8/yolov8.py:
```diff
@@ -0,0 +1,201 @@
+        run_name: str = None,
+        project: str = None,
+        tags: List[str] = None,
+        resume: str = "allow",
```
add `**kwargs`?

Commit:
feat(refactor): add top level import for ultralytics integration

Commit:
fix(integrations): add kwargs to logger and edit usage docstrings

TimH98 committed:
chore(launch): More detailed logging in runners (#4974)

* More detailed logging in runners

* Better wording for agent logging

fdsig committed:
fix(sdk): fix a typo in `CONTRIBUTING.md` (#4984)

fixing a typo in CONTRIBUTING.md

adding r to your in line 74 of CONTRIBUTING.md
  ``` 3.  Develop you(r) contribution. ```

Commit:
refactor(integrations): remove context manager and add functional wrapper instead

Based on the feedback from the SDK team it makes more sense to create a functional wrapper instead of a context manager.

This decision was due to the following reasons:
1. The context manager is more likely to fail in DDP processes.
2. In the event of an error occurring the error only appears once the context manager exits. This is not desirable.
3. The context manager also has the potential to confuse the user. A functional implementation makes it simpler

Commit:
feat(integrations):: add telemetry for tracking yolov8 feature

Commit:
feat(integrations):: add proto v4 telemetry for tracking yolov8 feature

dmitryduev committed:
fix(sdk): correctly detect notebook name and fix code saving in Colab (#4987)

fix(sdk): correctly detect notebook name and fix code saving in Colab

moredatarequired committed:
fix(artifacts): allow up to max_artifacts (fix off by 1 error) (#4991)

Fix off by 1 (allow up to max_objects, not max_objects - 1)

moredatarequired committed:
chore(sdk): Ignore wandb run data (#4952)

Ignore wandb run data

dmitryduev committed:
test(sdk): fix admin user creation and setting up passwords for ui for --wandb-debug (#4994)

test(sdk): fix admin user creation and setting up passwords for ui in --wandb-debug

kptkin committed:
style(launch): move launch errors closer to the code (#4995)

* move docker and launch tests out of main errors

* lint

kptkin committed:
test(sdk): create a shard for sweep and artifact system tests (#4993)

* split out main tests

* create different shard for sweep and artifacts

* fix codecov-check

* rename to sweeps

kptkin committed:
style(sdk): move mailbox error closer to the code (#4997)

dmitryduev committed:
chore(sdk): exercise extra caution when starting asset monitoring threads (#5007)

speezepearson committed:
test(sdk): unit tests for StepPrepare (#4979)

* test(sdk): unit tests for StepChecksum

* make test less hardcoded

* doc

* refactor queue wait time calculation

* lint

* increase sleep times in tests to reduce flakiness

* make test-sleeps EVEN LONGER :( :( :(

* lint: import types directly from internal_api, rather than qualifying them

* prefer time.monotonic to time.time

* refactor gather_batch out of StepPrepare; write a mock Queue impl for testing; add tests for finish/shutdown

* deflake test

* fix mypy errs, improve typing

---------

Co-authored-by: Hugh Wimberly <hugh.wimberly@wandb.com>

speezepearson committed:
test(artifacts): unit tests for WandbStoragePolicy.store_file (#4980)

* test(sdk): unit tests for StepChecksum

* make test less hardcoded

* doc

* refactor queue wait time calculation

* test(artifacts): unit tests for WandbStoragePolicy.store_file

* remove dead code

* fix return value of store_file helper-fn

* add assertions for store_file return value

* fix default value for mocked-out ArtifactManifestEntry.size: None, not 0

* lint

* lint

* increase sleep times in tests to reduce flakiness

* make test-sleeps EVEN LONGER :( :( :(

* Apply suggestions from code review

Co-authored-by: Dmitry Duev <dmitryduev@users.noreply.github.com>

* lint: import types directly from internal_api, rather than qualifying them

* prefer time.monotonic to time.time

* refactor gather_batch out of StepPrepare; write a mock Queue impl for testing; add tests for finish/shutdown

* deflake test

* fix mypy errs, improve typing

---------

Co-authored-by: Hugh Wimberly <hugh.wimberly@wandb.com>
Co-authored-by: Dmitry Duev <dmitryduev@users.noreply.github.com>

gtarpenning committed:
chore(launch): log better when default project  (#5010)

loglog

Commit:
test(integrations): add functional tests for ultralytics yolov8 integration

Commit:
chore(integrations): format with black and isort

Commit:
fix(integrations): remove wandb logger from module imports

Commit:
fix(integrations): change wandblogger to wandb callback

parambharat committed:
Merge branch 'wandb:main' into yolov8-integration

Commit:
test(integrations): increase test coverage and bump version of `ultralytics`

Commit:
Merge branch 'yolov8-integration' of github.com:parambharat/wandb into yolov8-integration

Commit:
test(integrations): fix test identification for segmentation

Commit:
fix(integrations): typos in docstrings and remove completed TODOs

parambharat committed:
Merge branch 'wandb:main' into yolov8-integration

Commit:
fix(integrations): show health warning only once in a run

Commit:
Merge branch 'yolov8-integration' of github.com:parambharat/wandb into yolov8-integration

Commit:
fix(integrations): validate inference speeds for different version keys

parambharat committed:
Merge branch 'wandb:main' into yolov8-integration

parambharat committed:
Merge branch 'main' into yolov8-integration

Commit:
chore(integrations): fix and update dev requirements for ultralytics

Commit:
Merge branch 'yolov8-integration' of github.com:wandb/wandb into yolov8-integration

parambharat committed:
Merge branch 'main' into yolov8-integration

dmitryduev committed:
Merge branch 'main' into yolov8-integration

Commit:
fix(integrations): change wandb resume default to `None`

fixed based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1123863566

Commit:
fix(integrations): update usage docs to not return wrapped object

fixed based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1123873438

Commit:
fix(integrations): remove compatibility module for new integration

fixed based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1123874081

Commit:
tests(integration): update tests to reflect changes made from review

- remove compatibility module and corresponding imports
- change model wrapper reference

Commit:
Merge branch 'yolov8-integration' of github.com:wandb/wandb into yolov8-integration

Commit:
Merge branch 'main' of github.com:wandb/wandb into yolov8-integration

Resolved conflicts in requirements and telemetry proto

# Conflicts:
#	requirements_dev.txt
#	wandb/proto/v3/wandb_telemetry_pb2.py
#	wandb/proto/v3/wandb_telemetry_pb2.pyi
#	wandb/proto/v4/wandb_telemetry_pb2.py
#	wandb/proto/v4/wandb_telemetry_pb2.pyi
#	wandb/proto/wandb_telemetry.proto

Change functional tests to reflect the changed telemetry proto

raubitsj committed:
fix up docstrings (part1)

raubitsj committed:
fix up docstrings (part2)

raubitsj committed:
Add mypy checking

Commit:
fix(integrations): remove `ultralytics` from dev requirements

based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1124995997

Commit:
fix(integrations): add type checking for return values

based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1124999008

Commit:
fix(integrations): explain and add defaults to docstrings

based on review comments:
 https://github.com/wandb/wandb/pull/5037#discussion_r1125000372
 https://github.com/wandb/wandb/pull/5037#discussion_r1125001623
 https://github.com/wandb/wandb/pull/5037#discussion_r1125002371
 https://github.com/wandb/wandb/pull/5037#discussion_r1125003050
 https://github.com/wandb/wandb/pull/5037#discussion_r1125011716

Commit:
fix(integrations): change the default for wandb resume to `None`

based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1125003784

Commit:
fix(integrations): explain why only detection and segmentation trainers log plots

based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1125007950

kptkin committed:
Merge branch 'main' into yolov8-integration

parambharat committed:
Merge branch 'main' into yolov8-integration

Commit:
fix(integrations): add typing annotations for mypy tests and sort imports

based on review comment: https://github.com/wandb/wandb/pull/5037#discussion_r1124999008
